### PR TITLE
support nested subquery and fix bug of setting variable

### PR DESCRIPTION
**Pull request by hanfei19910905**
@shenli @coocood @zimulala PTAL
1. support nested subquery, which is a subquery containing another subquery.
2. support queries like select @a:= b+1 from t;

### CODE REVIEW

File: expression/expression.go [hunk 1]
```diff
@@ -40,6 +41,32 @@ type Expression interface {
 
 	// ToString converts an expression into a string.
 	ToString() string
+
+	// SetVarName set values for session variable.
```
Comment (zimulala) [inline expression/expression.go]: set -> sets?

**Comment by shenli** [comment]
Could you describe the bug in conversation?

File: expression/expression.go [hunk 2]
```diff
@@ -40,6 +41,32 @@ type Expression interface {
...
+	SetVarName(string)
```
Comment (coocood) [inline expression/expression.go]: Set session variable is only used by a particular exprssion, we don't need to add it to Expression interface.

File: evaluator/builtin_other.go [hunk 3]
```diff
@@ -460,3 +462,27 @@ func CastFuncFactory(tp *types.FieldType) (BuiltinFunc, error) {
 	}
 	return nil, errors.Errorf("unknown cast type - %v", tp)
 }
+
+func builtinSetVar(args []types.Datum, ctx context.Context) (types.Datum, error) {
+	sessionVars := variable.GetSessionVars(ctx)
+	varName, _ := args[1].ToString()
```
Comment (coocood) [inline evaluator/builtin_other.go]: make varName args[0] is more intuitive.

File: evaluator/builtin_other.go [hunk 4]
```diff
@@ -460,3 +462,27 @@ func CastFuncFactory(tp *types.FieldType) (BuiltinFunc, error) {
 	}
 	return nil, errors.Errorf("unknown cast type - %v", tp)
 }
+
+func builtinSetVar(args []types.Datum, ctx context.Context) (types.Datum, error) {
+	sessionVars := variable.GetSessionVars(ctx)
+	varName, _ := args[0].ToString()
```
Comment (shenli) [inline evaluator/builtin_other.go]: Why ignore error?

File: evaluator/builtin_other.go [hunk 5]
```diff
@@ -460,3 +462,27 @@ func CastFuncFactory(tp *types.FieldType) (BuiltinFunc, error) {
...
+	if !args[1].IsNull() {
+		strVal, err := args[1].ToString()
+		if err != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+		sessionVars.Users[varName] = strings.ToLower(strVal)
```
Comment (shenli) [inline evaluator/builtin_other.go]: Will there be system vars?

File: evaluator/builtin_other.go [hunk 6]
```diff
@@ -460,3 +462,27 @@ func CastFuncFactory(tp *types.FieldType) (BuiltinFunc, error) {
...
+}
+
+func builtinGetVar(args []types.Datum, ctx context.Context) (types.Datum, error) {
+	sessionVars := variable.GetSessionVars(ctx)
+	varName, _ := args[0].ToString()
+	if v, ok := sessionVars.Users[varName]; ok {
+		return types.NewDatum(v), nil
+	}
+	return types.Datum{}, nil
+
```
Comment (shenli) [inline evaluator/builtin_other.go]: Remove the empty line. So as line 477.

File: evaluator/builtin_other.go (hunk 4 above)
Comment (hanfei19910905) [inline evaluator/builtin_other.go]: Because we set a string constant, it can't be wrong.

File: evaluator/builtin_other.go (hunk 5 above)
Comment (hanfei19910905) [inline evaluator/builtin_other.go]: Never.

**Comment by shenli** [comment]
LGTM

**Pull request merged**